consensus/src/txn_hash_and_authenticator_deduper.rs

```diff
@@ -0,0 +1,344 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    counters::{TXN_DEDUP_FILTERED, TXN_DEDUP_SECONDS},
+    transaction_deduper::TransactionDeduper,
+};
+use aptos_types::transaction::SignedTransaction;
+use rayon::prelude::*;
+use std::collections::{HashMap, HashSet};
+
+pub struct TxnHashAndAuthenticatorDeduper {}
+
+impl TransactionDeduper for TxnHashAndAuthenticatorDeduper {
+    fn dedup(&self, transactions: Vec<SignedTransaction>) -> Vec<SignedTransaction> {
+        let _timer = TXN_DEDUP_SECONDS.start_timer();
+        let mut seen = HashMap::new();
+        let mut is_possible_duplicate = false;
+        let mut possible_duplicates = vec![false; transactions.len()];
+        for (i, txn) in transactions.iter().enumerate() {
+            match seen.get(&(txn.sender(), txn.sequence_number())) {
+                None => {
+                    seen.insert((txn.sender(), txn.sequence_number()), i);
+                },
+                Some(first_index) => {
+                    is_possible_duplicate = true;
+                    possible_duplicates[*first_index] = true;
+                    possible_duplicates[i] = true;
+                },
+            }
+        }
+        if !is_possible_duplicate {
+            TXN_DEDUP_FILTERED.observe(0 as f64);
+            return transactions;
+        }
+
+        let hash_and_authenticators: Vec<_> = possible_duplicates
+            .into_par_iter()
+            .zip(&transactions)
+            .with_min_len(25)
+            .map(|(need_hash, txn)| match need_hash {
```

Comment:
can we use filter_map here? would remove all the None-s and eliminate Option wrapping.

Reply:
but we just post-process it to duplicates and then zip it with transactions (true/false), which filters, but instead we could also add txn here (maybe into_iter), and then we shouldn't need to zip at the end of the function?

Reply:
I combined the final two steps (filter + map) into a single step (filter_map). The comment above was on the second step, when duplicates aren't completely figured out yet, so I don't think we can combine that step.